ui/jobs/jobs.js

```diff
@@ -1726,11 +1789,134 @@ class Bars {
+      this.museStacks = ethosStacks ? ethosStacks : this.paeonStacks;
+      this.paeonStacks = 0;
+    };
+    this.loseEffectFuncMap[gLang.kEffect.ArmysMuse] = () => {
+      // Muse effect ends
+      this.museStacks = 0;
+      this.paeonStacks = 0;
+    };
+    this.gainEffectFuncMap[gLang.kEffect.ArmysEthos] = (name, log) => {
+      // Not under muse or paeon, so store the stacks
+      ethosStacks = this.paeonStacks;
+      this.paeonStacks = 0;
+    };
+    this.loseEffectFuncMap[gLang.kEffect.ArmysEthos] = () => {
+      // Didn't use a song and ethos ran out
+      ethosStacks = 0;
+      this.museStacks = 0;
+      this.paeonStacks = 0;
+    };
+
+    this.jobFuncs.push((jobDetail) => {
+      if (jobDetail.songName == 'Paeon' && this.paeonStacks != jobDetail.songProcs)
+        this.paeonStacks = jobDetail.songProcs;
+    });
+  }
+
+  setupWhm() {
+    this.gainEffectFuncMap[gLang.kEffect.PresenceOfMind] = (name, log) => {
+      this.presenceOfMind = 1;
+    };
+    this.loseEffectFuncMap[gLang.kEffect.PresenceOfMind] = () => {
+      this.presenceOfMind = 0;
+    };
+  }
+
+  setupNin() {
+    this.jobFuncs.push((jobDetail) => {
+      if (jobDetail.hutonMilliseconds > 0) {
+        if (this.huton != 1)
+          this.huton = 1;
+      } else if (this.huton == 1) {
+        this.huton = 0;
+      }
+    });
+  }
+
+  setupSam() {
+    this.gainEffectFuncMap[gLang.kEffect.Shifu] = (name, log) => {
+      this.shifu = 1;
+    };
+    this.loseEffectFuncMap[gLang.kEffect.Shifu] = () => {
+      this.shifu = 0;
+    };
+  }
+
   OnComboChange(skill) {
     for (let i = 0; i < this.comboFuncs.length; ++i)
       this.comboFuncs[i](skill);
   }
 
+  // Source: http://theoryjerks.akhmorning.com/guide/speed/
+  CalcGCDFromStat(stat, actiondelay) {
+    // default calculates for a 2.50s recast
+    actiondelay = actiondelay || 2500;
+
+    let type1Buffs = 0;
+    let type2Buffs = 0;
+    if (this.job == 'BLM') {
+      type1Buffs += this.circleOfPower ? 15 : 0;
+    } else if (this.job == 'WHM') {
+      type1Buffs += this.presenceOfMind ? 20 : 0;
+    } else if (this.job == 'SAM') {
+      if (this.shifu) {
+        if (this.level > 77)
+          type1Buffs += 13;
+        else type1Buffs += 10;
+      }
+    }
+
+    if (this.job == 'NIN') {
+      type2Buffs += this.huton ? 15 : 0;
+    } else if (this.job == 'MNK') {
+      type2Buffs += 5 * this.lightningStacks;
+    } else if (this.job == 'BRD') {
+      type2Buffs += 4 * this.paeonStacks;
+      switch (this.museStacks) {
+      case 1:
+        type2Buffs += 1;
+        break;
+      case 2:
+        type2Buffs += 2;
+        break;
+      case 3:
+        type2Buffs += 4;
+        break;
+      case 4:
+        type2Buffs += 12;
+        break;
+      }
+    }
+    // TODO: this probably isn't useful to track
+    let astralUmbralMod = 100;
+
+    let GCDms = Math.floor(1000 - Math.floor(130 * (stat - kLevelMod[this.level][0]) /
```

Comment:
It felt neater that you could just pass `this.level` rather than subtracting one so I bumped the `kLevelMod` entries to match.

Reply:
:+1: